### PR TITLE
Feature/ecu battery module

### PR DESCRIPTION
**Pull request by RubenNatanael**
Creation of modules for:
    -Battery
    -Interface creation&configuration
    -Receive and generate frames via UDS protocol.
    -Unit tests for the modules

### CODE REVIEW

File: src/ecu_simulation/BatteryModule/include/GenerateFrames.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        void accessTimingParameters(int id, int sub_function, bool response=false);
+        void negativeResponse(int id, int nrc);
+        //OTA
+        void requestDownload(int id, int data_format_identifier, int memory_address, int memory_size);
+        void requestDownloadResponse(int id, int max_number_block);
+        void transferData(int id, int block_sequence_counter, std::vector<int> transfer_request = {});
+        void transferDataLong(int id, int block_sequence_counter, std::vector<int> transfer_request = {}, bool first_frame = true);
+        void requestTransferExit(int id, bool response=false);
+        bool requestUpdateStatus(int id, bool response=false);
+    private:
```
Comment (GeorgeCimpoeru) [inline src/ecu_simulation/BatteryModule/include/GenerateFrames.h:70]: why do we have 2 private sectors in this class ?

**Comment by GeorgeCimpoeru** [comment]
For all : 
Please check your comments and establish with all the teams on which variant are you going : 
1 . using "//"
2 . using "/*   */"

Personally I would go with "/* */" . It is more visible and more clear to read.

Also, Please add descriptions for all methods inside .h files (description should contain all necessary information from what is the method doing, to how many parameters are requested and what is returning).

The necessary description should be available inside .h files.
In the .cpp files, we should also add description, but very short on what exactly is supposed to do that function/method.

File: src/ecu_simulation/BatteryModule/include/ReceiveFrames.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    bool running; // Flag to indicate when to stop the threads
+    std::thread producerThread;
+    std::thread consumerThread;
+
+    void Producer();
+    void Consumer(HandleFrames &handleFrame);
+protected:
+    HandleFrames handleFrame;
+    
+
```
Comment (GeorgeCimpoeru) [inline src/ecu_simulation/BatteryModule/include/ReceiveFrames.h:31]: Please remove this empty line

File: src/ecu_simulation/BatteryModule/src/HandleFrames.cpp [hunk 3]
```diff
@@ -0,0 +1,254 @@
+/*Author:Stoisor Miruna, 2024*/
```
Comment (GeorgeCimpoeru) [inline src/ecu_simulation/BatteryModule/src/HandleFrames.cpp:1]: Please establish with the rest of the team and set a wide description of the current file.

File: src/ecu_simulation/BatteryModule/src/HandleFrames.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+                std::cout << "Valid data length\n"<<std::flush;
+                //SessionControl(id, sub_function);
+            } else {
+                std::cerr << "Invalid data length for SessionControl\n";
+            }  
+            exit(0);        
+            break;
+        case 0x11: // EcuReset
+            std::cout << "Service 0x11 EcuReset\n";
+            if (sid_position + 1 < data.size()) {  //sub_function data always found next to sid, has to be in valid data range
```
Comment (GeorgeCimpoeru) [inline src/ecu_simulation/BatteryModule/src/HandleFrames.cpp:57]: please avoid to append comments at the end of line.
Add the proper comments before the line.

File: src/ecu_simulation/BatteryModule/src/HandleFrames.cpp [hunk 5]
```diff
@@ -0,0 +1,254 @@
+/*Author:Stoisor Miruna, 2024*/
+
+#include "../include/HandleFrames.h"
+
+
+/*Services to be included here*/
+
+void HandleFrames::ProcessReceivedFrame(const struct can_frame &frame) {
```
Comment (GeorgeCimpoeru) [inline src/ecu_simulation/BatteryModule/src/HandleFrames.cpp:8]: Please check with the rest of the teams and establish together if you will use  : 
"{" at the end of the line or on the next line.
example : 
1. if (isFirstFrame) {
2. if (isFirstFrame) 
   {

File: src/ecu_simulation/BatteryModule/src/InterfaceConfig.cpp [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+ * 
+ */
+
+#include "../include/InterfaceConfig.h"
+#include <iostream> 
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+#include <linux/can/raw.h>
+#include <string.h>
+#include <unistd.h>
```
Comment (GeorgeCimpoeru) [inline src/ecu_simulation/BatteryModule/src/InterfaceConfig.cpp:18]: Please add the other includes inside the .h file correspondent with our cpp file

File: src/ecu_simulation/BatteryModule/src/ReceiveFrames.cpp [hunk 7]
```diff
@@ -0,0 +1,126 @@
+/*Author:Stoisor Miruna, 2024*/
+
+#include <sstream>
+#include <iomanip>
+#include <cstring>
```
Comment (GeorgeCimpoeru) [inline src/ecu_simulation/BatteryModule/src/ReceiveFrames.cpp:5]: Please add the includes of other libraries inside the .h file

File: src/ecu_simulation/BatteryModule/src/main.cpp [hunk 8]
```diff
@@ -0,0 +1,12 @@
+#include <iostream>
+#include <thread>
```
Comment (GeorgeCimpoeru) [inline src/ecu_simulation/BatteryModule/src/main.cpp:2]: please add these includes inside the .h file

File: src/ecu_simulation/BatteryModule/.gitignore [hunk 9]
```diff
@@ -0,0 +1,5 @@
+test_local*
+/src/Generate.cpp
+/src/Receive.cpp
+teste*
+/test/build*
```
[the remark this answers is not in the document]
Comment (GeorgeCimpoeru) [reply src/ecu_simulation/BatteryModule/.gitignore:5]: ok, then this one can stay

File: src/ecu_simulation/BatteryModule/src/InterfaceConfig.cpp (hunk 6 above)
Comment (IDaniel99) [reply src/ecu_simulation/BatteryModule/src/InterfaceConfig.cpp]: done

File: src/ecu_simulation/BatteryModule/include/InterfaceConfig.h [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    void createLinuxVCanInterface();
+    void deleteLinuxVCanInterface();
+
+    bool openInterface();
+    void closeInterface();
+
+public:
+    /* Init vcan communication */
+
+    /* Create vcan  */
```
[the remark this answers is not in the document]
Comment (IDaniel99) [reply src/ecu_simulation/BatteryModule/include/InterfaceConfig.h]: done

File: src/ecu_simulation/BatteryModule/src/BatteryModule.cpp [hunk 11]
```diff
@@ -0,0 +1,141 @@
+#include "../include/BatteryModule.h"
+#include <cstdlib>
+#include <chrono>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (nagyyalexandru) [reply src/ecu_simulation/BatteryModule/src/BatteryModule.cpp]: ok, it's done

File: src/ecu_simulation/BatteryModule/src/ReceiveFrames.cpp
Comment (GeorgeCimpoeru) [inline src/ecu_simulation/BatteryModule/src/ReceiveFrames.cpp:1]: Please rename the Consumer and Producer to be more explicit.

**Review by GeorgeCimpoeru — APPROVED** [review]

**Review by vlcristi91 — APPROVED** [review]

**Pull request merged**